File: apps/macos/VoxtralRealtimeApp/VoxtralRealtime/Services/RunnerBridge.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import os

private let log = Logger(subsystem: "org.pytorch.executorch.VoxtralRealtime", category: "RunnerBridge")

actor RunnerBridge {
    enum ModelState: Sendable, Equatable {
        case unloaded
        case loading
        case ready
    }

    struct Streams: Sendable {
        let tokens: AsyncStream<String>
        let errors: AsyncStream<RunnerError>
        let modelState: AsyncStream<ModelState>
        let audioLevel: AsyncStream<Float>
        let status: AsyncStream<String>
    }

    private var process: Process?
    private var stdinPipe: Pipe?
    private let audioEngine = AudioEngine()
    private var levelContinuation: AsyncStream<Float>.Continuation?
    private var tokenContinuation: AsyncStream<String>.Continuation?
    private var statusContinuation: AsyncStream<String>.Continuation?
    private var errorContinuation: AsyncStream<RunnerError>.Continuation?
    private var modelStateContinuation: AsyncStream<ModelState>.Continuation?

    private(set) var modelState: ModelState = .unloaded

    var isRunnerAlive: Bool { process?.isRunning == true }

    // MARK: - Process lifecycle

    func launchRunner(
        runnerPath: String,
        modelPath: String,
        tokenizerPath: String,
        preprocessorPath: String
    ) -> Streams {
        if isRunnerAlive { stop() }

        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        let stdinPipe = Pipe()
        self.stdinPipe = stdinPipe

        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: runnerPath)
        proc.arguments = [
            "--model_path", modelPath,
            "--tokenizer_path", tokenizerPath,
            "--preprocessor_path", preprocessorPath,
            "--mic"
        ]

        var env = ProcessInfo.processInfo.environment
        if let bundleResources = Bundle.main.resourcePath {
            let existing = env["DYLD_LIBRARY_PATH"] ?? ""
            env["DYLD_LIBRARY_PATH"] = existing.isEmpty ? bundleResources : "\(bundleResources):\(existing)"
        }
        proc.environment = env
        proc.standardInput = stdinPipe
        proc.standardOutput = stdoutPipe
        proc.standardError = stderrPipe
        self.process = proc

        let (tokenStream, tokenCont) = AsyncStream<String>.makeStream()
        let (errorStream, errorCont) = AsyncStream<RunnerError>.makeStream()
        let (modelStateStream, modelStateCont) = AsyncStream<ModelState>.makeStream()
        let (levelStream, levelCont) = AsyncStream<Float>.makeStream()
        let (statusStream, statusCont) = AsyncStream<String>.makeStream()

        self.tokenContinuation = tokenCont
        self.errorContinuation = errorCont
        self.modelStateContinuation = modelStateCont
        self.levelContinuation = levelCont
        self.statusContinuation = statusCont

        modelState = .loading
        modelStateCont.yield(.loading)

        log.info("Launching runner: \(runnerPath)")
        statusCont.yield("Launching runner...")

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let handle = stdoutPipe.fileHandleForReading
            var sawListening = false

            while true {
                let data = handle.availableData
                if data.isEmpty {
                    log.info("Runner stdout EOF")
                    tokenCont.finish()
                    break
                }

                guard let text = String(data: data, encoding: .utf8) else { continue }
                log.debug("Runner stdout (\(data.count) bytes): \(text.prefix(200))")

                if !sawListening && text.contains("Listening") {
                    sawListening = true
                    log.info("Model loaded — runner ready")
                    modelStateCont.yield(.ready)
                    Task { await self?.setModelState(.ready) }

                    let remainder = text.replacingOccurrences(of: "Listening (Ctrl+C to stop)...", with: "")
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                    if !remainder.isEmpty { tokenCont.yield(remainder) }
                    continue
                }

                if text.contains("PyTorchObserver") {
                    let parts = text.components(separatedBy: "\n")
                    let nonStats = parts.filter { !$0.contains("PyTorchObserver") }.joined(separator: "\n")
                    if !nonStats.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                        tokenCont.yield(nonStats)
                    }
                    continue
                }

                let cleaned = text.replacingOccurrences(
                    of: "\u{1B}\\[[0-9;]*m", with: "", options: .regularExpression
                )
                if !cleaned.isEmpty { tokenCont.yield(cleaned) }
            }
        }

        DispatchQueue.global(qos: .utility).async {
            let handle = stderrPipe.fileHandleForReading
            while true {
                let data = handle.availableData
                if data.isEmpty { break }
                guard let text = String(data: data, encoding: .utf8) else { continue }
                log.debug("Runner stderr: \(text.trimmingCharacters(in: .newlines))")

                let lastLine = text.components(separatedBy: "\n")
                    .filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
                    .last ?? ""
                if lastLine.contains("Loading model") { statusCont.yield("Loading model...") }
                else if lastLine.contains("Loading tokenizer") { statusCont.yield("Loading tokenizer...") }
                else if lastLine.contains("Loading preprocessor") { statusCont.yield("Loading preprocessor...") }
                else if lastLine.contains("Warming up") { statusCont.yield("Warming up...") }
                else if lastLine.contains("Warmup complete") { statusCont.yield("Model ready") }
            }
        }

        proc.terminationHandler = { [weak self] process in
            let code = process.terminationStatus
            log.info("Runner exited with code \(code)")
            if code != 0 && code != 2 {
                errorCont.yield(.runnerCrashed(exitCode: code, stderr: "Exit code: \(code)"))
            }
            tokenCont.finish()
            errorCont.finish()
            levelCont.finish()
            statusCont.finish()
            modelStateCont.yield(.unloaded)
            modelStateCont.finish()
            Task { await self?.onProcessExited() }
        }

        do {
            try proc.run()
            log.info("Runner process started (pid: \(proc.processIdentifier))")
        } catch {
            log.error("Failed to launch runner: \(error.localizedDescription)")
            errorCont.yield(.launchFailed(description: error.localizedDescription))
            errorCont.finish()
            tokenCont.finish()
            levelCont.finish()
            statusCont.finish()
            modelStateCont.yield(.unloaded)
            modelStateCont.finish()
        }

        return Streams(
            tokens: tokenStream, errors: errorStream, modelState: modelStateStream,
            audioLevel: levelStream, status: statusStream
        )
    }

    // MARK: - Audio lifecycle (independent of process)

    func startAudioCapture() async throws {
        guard let handle = stdinPipe?.fileHandleForWriting else {
            throw RunnerError.launchFailed(description: "Runner stdin not available")
        }
        let cont = self.levelContinuation
        log.info("Starting audio capture")
        try await audioEngine.startCapture(writingTo: handle) { level in
            cont?.yield(level)
        }
    }

    func stopAudioCapture() async {
        await audioEngine.stopCapture()
        levelContinuation?.yield(0)
        log.info("Audio capture stopped (runner still alive)")
    }

    // MARK: - Full shutdown

    func stop() {
        Task { await audioEngine.stopCapture() }

        stdinPipe?.fileHandleForWriting.closeFile()
        stdinPipe = nil

        if let proc = process, proc.isRunning {
            proc.interrupt()
            proc.waitUntilExit()
        }
        process = nil
        modelState = .unloaded
        clearContinuations()
    }

    // MARK: - Private

    private func setModelState(_ state: ModelState) {
        modelState = state
    }

    private func onProcessExited() {
        stdinPipe = nil
        process = nil
        modelState = .unloaded
        clearContinuations()
    }

    private func clearContinuations() {
        tokenContinuation = nil
        errorContinuation = nil
        modelStateContinuation = nil
        levelContinuation = nil
        statusContinuation = nil
    }
}

File: voxtral_realtime/windows/VoxtralRealtime/VoxtralRealtime/Services/RunnerBridge.cs
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

using System.Diagnostics;
using System.IO;
using System.Text.RegularExpressions;
using VoxtralRealtime.Models;

namespace VoxtralRealtime.Services;

public class RunnerBridge : IDisposable
{
    private Process? _process;
    // Hold strong reference to StandardInput to prevent GC from closing the pipe
    private StreamWriter? _stdinWriter;
    private Stream? _stdinStream;
    private AudioCaptureService? _audioCapture;
    private readonly object _lock = new();
    private bool _disposed;

    public event Action<string>? TokenReceived;
    public event Action<string>? StatusReceived;
    public event Action<string>? ErrorOccurred;
    public event Action<ModelState>? ModelStateChanged;
    public event Action<float>? AudioLevelChanged;

    public ModelState CurrentModelState { get; private set; } = ModelState.Unloaded;
    public bool IsRunnerAlive
    {
        get
        {
            lock (_lock)
            {
                return _process is { HasExited: false };
            }
        }
    }

    private static readonly Regex AnsiRegex = new(@"\x1B\[[0-9;]*m", RegexOptions.Compiled);

    public void LaunchRunner(
        string runnerPath,
        string modelPath,
        string tokenizerPath,
        string preprocessorPath,
        string dataPath)
    {
        lock (_lock)
        {
            if (IsRunnerAlive) Stop();
        }

        var runnerDir = Path.GetDirectoryName(runnerPath) ?? "";
        var args = BuildArguments(modelPath, tokenizerPath, preprocessorPath, dataPath);
        AppLogger.Log("Runner", $"Launching: {runnerPath}");
        AppLogger.Log("Runner", $"Args: {args}");

        var startInfo = new ProcessStartInfo
        {
            FileName = runnerPath,
            Arguments = args,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };

        // Set PATH to include runner directory and CUDA toolkit
        var extraPaths = new List<string> { runnerDir };
        extraPaths.AddRange(FindCudaPaths());
        var existingPath = Environment.GetEnvironmentVariable("PATH") ?? "";
        startInfo.Environment["PATH"] = string.Join(";", extraPaths) + ";" + existingPath;
        AppLogger.Log("Runner", $"Extra PATH dirs: {string.Join("; ", extraPaths)}");

        var proc = new Process { StartInfo = startInfo, EnableRaisingEvents = true };

        proc.Exited += (_, _) =>
        {
            var exitCode = proc.ExitCode;
            AppLogger.Log("Runner", $"Runner EXITED with code {exitCode} (0x{exitCode:X8})");
            if (exitCode != 0 && exitCode != 2)
            {
                ErrorOccurred?.Invoke($"Runner crashed with exit code {exitCode}");
            }
            CurrentModelState = ModelState.Unloaded;
            ModelStateChanged?.Invoke(ModelState.Unloaded);
        };

        lock (_lock) { _process = proc; }

        CurrentModelState = ModelState.Loading;
        ModelStateChanged?.Invoke(ModelState.Loading);
        StatusReceived?.Invoke("Launching runner...");

        try
        {
            proc.Start();
            AppLogger.Log("Runner", $"Runner started (PID: {proc.Id})");

            // CRITICAL: Hold a strong reference to StandardInput (StreamWriter)
            // to prevent GC from finalizing it and closing the underlying pipe.
            // Also grab the raw BaseStream for binary audio writes.
            _stdinWriter = proc.StandardInput;
            _stdinWriter.AutoFlush = false;
            _stdinStream = _stdinWriter.BaseStream;
            AppLogger.Log("Runner", "Stdin pipe acquired");

            // Pre-warm audio device while model loads (takes ~7s, overlaps with model load)
            if (_audioCapture == null)
            {
                _audioCapture = new AudioCaptureService();
                _audioCapture.LevelChanged += level => AudioLevelChanged?.Invoke(level);
                Task.Run(() =>
                {
                    try
                    {
                        _audioCapture.InitAndStartDevice();
                        AppLogger.Log("Runner", "Audio device pre-warmed");
                    }
                    catch (Exception ex)
                    {
                        AppLogger.Log("Runner", $"Audio pre-warm failed: {ex.Message}");
                    }
                });
            }
        }
        catch (Exception ex)
        {
            AppLogger.Log("Runner", $"Failed to launch: {ex.Message}");
            ErrorOccurred?.Invoke($"Failed to launch runner: {ex.Message}");
            CurrentModelState = ModelState.Unloaded;
            ModelStateChanged?.Invoke(ModelState.Unloaded);
            return;
        }

        Task.Run(() => ReadStdout(proc));
        Task.Run(() => ReadStderr(proc));
    }

    private static string BuildArguments(
        string modelPath, string tokenizerPath,
        string preprocessorPath, string dataPath)
    {
        var args = $"--model_path \"{modelPath}\" " +
                   $"--tokenizer_path \"{tokenizerPath}\" " +
                   $"--preprocessor_path \"{preprocessorPath}\"";

        if (!string.IsNullOrEmpty(dataPath))
        {
            args += $" --data_path \"{dataPath}\"";
        }

        args += " --mic";
        return args;
    }

    private void ReadStdout(Process proc)
    {
        var stream = proc.StandardOutput;
        bool sawListening = false;
        var buffer = new char[4096];
        var lineBuffer = new System.Text.StringBuilder();

        try
        {
            while (true)
            {
                int charsRead = stream.Read(buffer, 0, buffer.Length);
                if (charsRead == 0) break;

                var text = new string(buffer, 0, charsRead);
                AppLogger.Log("Stdout", $"({charsRead} chars): {text[..Math.Min(300, text.Length)]}");

                if (!sawListening)
                {
                    lineBuffer.Append(text);
                    var accumulated = lineBuffer.ToString();
                    if (accumulated.Contains("Listening"))
                    {
                        sawListening = true;
                        CurrentModelState = ModelState.Ready;
                        ModelStateChanged?.Invoke(ModelState.Ready);
                        AppLogger.Log("Runner", "Model ready - saw 'Listening'");

                        var remainder = accumulated
                            .Replace("Listening (Ctrl+C to stop)...", "")
                            .Trim();
                        if (!string.IsNullOrEmpty(remainder))
                        {
                            var cleaned = AnsiRegex.Replace(remainder, "");
                            if (!string.IsNullOrEmpty(cleaned))
                                TokenReceived?.Invoke(cleaned);
                        }
                        lineBuffer.Clear();
                    }
                    continue;
                }

                if (text.Contains("PyTorchObserver"))
                {
                    var parts = text.Split('\n');
                    var nonStats = string.Join("", parts.Where(p => !p.Contains("PyTorchObserver")));
                    if (!string.IsNullOrWhiteSpace(nonStats))
                    {
                        var cleaned = AnsiRegex.Replace(nonStats, "");
                        if (!string.IsNullOrEmpty(cleaned))
                            TokenReceived?.Invoke(cleaned);
                    }
                    continue;
                }

                var cleanedText = AnsiRegex.Replace(text, "");
                if (!string.IsNullOrEmpty(cleanedText))
                {
                    TokenReceived?.Invoke(cleanedText);
                }
            }
        }
        catch (Exception ex)
        {
            AppLogger.Log("Stdout", $"Reader error: {ex.Message}");
        }
        AppLogger.Log("Stdout", "Reader thread exiting");
    }

    private void ReadStderr(Process proc)
    {
        var reader = proc.StandardError;
        try
        {
            while (!proc.HasExited || !reader.EndOfStream)
            {
                var line = reader.ReadLine();
                if (line == null) break;

                AppLogger.Log("Stderr", line);

                if (line.Contains("Loading model"))
                    StatusReceived?.Invoke("Loading model...");
                else if (line.Contains("Loading tokenizer"))
                    StatusReceived?.Invoke("Loading tokenizer...");
                else if (line.Contains("Loading preprocessor"))
                    StatusReceived?.Invoke("Loading preprocessor...");
                else if (line.Contains("Warming up"))
                    StatusReceived?.Invoke("Warming up...");
                else if (line.Contains("Warmup complete"))
                    StatusReceived?.Invoke("Model ready");
            }
        }
        catch (Exception ex)
        {
            AppLogger.Log("Stderr", $"Reader error: {ex.Message}");
        }
        AppLogger.Log("Stderr", "Reader thread exiting");
    }

    public void StartAudioCapture()
    {
        Process? proc;
        lock (_lock) { proc = _process; }

        if (proc == null || proc.HasExited)
        {
            AppLogger.Log("Runner", "StartAudioCapture FAILED: runner not alive");
            throw new InvalidOperationException("Runner is not running");
        }

        if (_stdinStream == null || !_stdinStream.CanWrite)
        {
            AppLogger.Log("Runner", $"StartAudioCapture FAILED: stdin null={_stdinStream == null}, canWrite={_stdinStream?.CanWrite}");
            throw new InvalidOperationException("Runner stdin not available");
        }

        AppLogger.Log("Runner", $"Starting audio capture, runner PID={proc.Id}");

        if (_audioCapture == null)
        {
            // Audio wasn't pre-warmed, initialize now
            _audioCapture = new AudioCaptureService();
            _audioCapture.LevelChanged += level => AudioLevelChanged?.Invoke(level);
            _audioCapture.InitAndStartDevice();
        }

        _audioCapture.StartPiping(_stdinStream);
        AppLogger.Log("Runner", "Audio piping enabled (instant)");
    }

    public void PrimeAudioSamples(float[] samples)
    {
        if (samples.Length == 0) return;
        if (_stdinStream == null)
            throw new InvalidOperationException("Runner stdin not available");

        var bytes = new byte[samples.Length * sizeof(float)];
        Buffer.BlockCopy(samples, 0, bytes, 0, bytes.Length);
        _stdinStream.Write(bytes, 0, bytes.Length);
    }

    public void StopAudioCapture()
    {
        AppLogger.Log("Runner", "Stopping audio capture");
        _audioCapture?.StopPiping();
        AudioLevelChanged?.Invoke(0);
    }

    public void Stop()
    {
        AppLogger.Log("Runner", "Stop() called");
        _audioCapture?.Dispose();
        _audioCapture = null;
        AudioLevelChanged?.Invoke(0);

        Process? proc;
        lock (_lock)
        {
            proc = _process;
            _process = null;
        }

        // Close stdin to signal runner to exit
        try
        {
            _stdinWriter?.Close();
            AppLogger.Log("Runner", "Stdin closed");
        }
        catch { }
        _stdinWriter = null;
        _stdinStream = null;

        if (proc == null) return;

        try
        {
            if (!proc.HasExited)
            {
                proc.WaitForExit(3000);
                if (!proc.HasExited)
                {
                    proc.Kill();
                    AppLogger.Log("Runner", "Runner killed");
                }
            }
        }
        catch { }

        proc.Dispose();
        CurrentModelState = ModelState.Unloaded;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        Stop();
    }

    private static List<string> FindCudaPaths()
    {
        var paths = new List<string>();

        var cudaPath = Environment.GetEnvironmentVariable("CUDA_PATH");
        if (!string.IsNullOrEmpty(cudaPath))
        {
            var bin = Path.Combine(cudaPath, "bin");
            if (Directory.Exists(bin)) paths.Add(bin);
        }

        var toolkitRoot = @"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA";
        if (Directory.Exists(toolkitRoot))
        {
            foreach (var dir in Directory.GetDirectories(toolkitRoot)
                         .OrderByDescending(d => d))
            {
                var bin = Path.Combine(dir, "bin");
                if (Directory.Exists(bin) && !paths.Contains(bin))
                    paths.Add(bin);
            }
        }

        var condaPrefix = Environment.GetEnvironmentVariable("CONDA_PREFIX");
        if (!string.IsNullOrEmpty(condaPrefix))
        {
            var libBin = Path.Combine(condaPrefix, "Library", "bin");
            if (Directory.Exists(libBin)) paths.Add(libBin);
            var bin = Path.Combine(condaPrefix, "bin");
            if (Directory.Exists(bin)) paths.Add(bin);
        }

        return paths;
    }
}
